### DISCUSSION TITLE
[Chapter 02] spring-boot-autoconfigration?

### DISCUSSION THREAD

**Discussion by ahngilwoong** [cowaine/spring-boot-for-msa#8] [Chapter 02. Framework and Spring Boot] [answered]
- 책에서 autoconfigration에 대한 얘기가 나옵니다.
  직역하자면 자동 설정이고 @SpringBootApplication 안의 @EnableAutoConfigration이 부트의 자동 설정을 담당한다고 하는데
  이 어노테이션의 자세한 역할에 대해 알아보고 싶습니다.

**Comment by corock** [comment, marked as answer]
`@EnableAutoConfigration` 애너테이션은 스프링 부트에서 자동 구성 설정을 사용하겠다는 일종의 선언입니다.
주목할 점은 컴포넌트 스캔을 먼저 진행하고, 자동 구성 설정(_auto configuration_)을 하는 건데요.

기본 설정은 `spring-boot-autoconfigure-x.x.x/META-INF/spring.factories` 의존성 내 파일에 정의된 구성 값들을 자동으로 등록합니다.

_e.g. spring-boot-autoconfigure-2.7.5/META-INF/spring.factories_

```properties
# Auto Configuration Import Listeners
org.springframework.boot.autoconfigure.AutoConfigurationImportListener=\
org.springframework.boot.autoconfigure.condition.ConditionEvaluationReportAutoConfigurationImportListener

# Auto Configuration Import Filters
org.springframework.boot.autoconfigure.AutoConfigurationImportFilter=\
org.springframework.boot.autoconfigure.condition.OnBeanCondition,\
org.springframework.boot.autoconfigure.condition.OnClassCondition,\
org.springframework.boot.autoconfigure.condition.OnWebApplicationCondition

# Auto Configure
# (...)
```

만약 자동 구성 설정을 제외하고 싶은 구성 빈이 있을 경우 `@EnableAutoConfigration` 애너테이션의 `exclude()` 메서드를 사용합니다.

```java
@SpringBootApplication(exclude = RedisAutoConfiguration.class)
public class CorockApplication {

    public static void main(String[] args) {
        SpringApplication.run(CorockApplication.class, args);
    }

}
```

부가로 IntelliJ IDEA 에서는 별도의 설정 없이 Actuator 탭에서 어떤 구성 빈이 등록되었는지 쉽게 확인할 수 있다고 하네요.
(_cf._ [SpringBoot AutoConfiguration을 대하는 자세](https://tecoble.techcourse.co.kr/post/2021-10-14-springboot-autoconfiguration/))
